jdaviz/configs/mosviz/plugins/parsers.py

```diff
@@ -379,37 +380,84 @@ def mos_niriss_parser(app, data_dir, obs_label=""):
-                  "Direct Image": direct_image,
-                  "2D Spectra C": spec2d_c,
-                  "2D Spectra R": spec2d_r,
-                  "1D Spectra C": spec1d_c,
-                  "1D Spectra R": spec1d_r
-                 }
-
-    # Convert from pathlib Paths back to strings
-    for key in file_lists:
-        file_lists[key] = [str(x) for x in file_lists[key]]
+
+    file_types = ['Source Catalog', 'Direct Image', '2D Spectra C',
+                  '2D Spectra R', '1D Spectra C', '1D Spectra R']
+    file_lists = {k: [] for k in file_types}
+
+    # use FITS header keywords to sort the directory's files
+    for filepath in path.glob('*'):
+        if filepath.is_dir():
+            continue
+
+        filestr = str(filepath)
+
+        if filepath.suffix == '.fits':
+            # eligible files will have a DATAMODL value in their primary headers
+            header = fits.getheader(filepath, ext=0)
+            datamodl = header.get('DATAMODL')
+
+            # infer the dispersion direction of 1D and 2D spectra by the last
+            # letter of their filter values (e.g., "GR150C" or "GR150R")
+            try:
+                dispersion = header.get('FILTER')[-1]
+            except TypeError:
+                dispersion = None # could be cleaner...
+
+            # distinguish image files from counts files -- only the former go
+            # through the "Assign World Coordinate System" calibration step
+            s_wcs = header.get('S_WCS')
+
+            if datamodl is None:
+                continue
+            elif datamodl == 'MultiSpecModel' and dispersion == 'C':
+                file_lists['1D Spectra C'].append(filestr)
+            elif datamodl == 'MultiSpecModel' and dispersion == 'R':
+                file_lists['1D Spectra R'].append(filestr)
+            elif datamodl == 'MultiSlitModel' and dispersion == 'C':
+                file_lists['2D Spectra C'].append(filestr)
+            elif datamodl == 'MultiSlitModel' and dispersion == 'R':
+                file_lists['2D Spectra R'].append(filestr)
+            elif datamodl == 'ImageModel' and s_wcs is not None:
+                file_lists['Direct Image'].append(filestr)
+            else:
+                # print(f"Other case: datamodl is {datamodl}; "
+                #       f"dispersion is {dispersion}")
+                pass
+
+        elif filepath.suffix == '.ecsv':
+            file_lists['Source Catalog'].append(filestr)
+
+        else:
+            continue
 
     _warn_if_not_found(app, file_lists)
 
     # Parse relevant information from source catalog
-    cat_fields = ["id", "sky_centroid.ra", "sky_centroid.dec"]
+    # TEMP NOTE: the ecsv table in the example directory has one column with
+    # sky centroid info given as an RA/Dec tuple instead of two columns with
+    # separate floats. i tried to support both schemas.
+    cat_fields = ['id', 'sky_centroid']
+    if 'sky_centroid' not in file_lists['Source Catalog'][0]:
```

Comment:
I am a bit lost here. `file_lists['Source Catalog']` is still a list of filepaths, right? Why would column name be in it? Don't you have to parse the actual file into `astropy.table.Table` first? Did I miss something?

`pop` and `extend` is a little confusing, I would prefer a simpler if-else using `+` on list, but that is just a matter of personal taste.

Reply:
You're right; this is my mistake. (The pop and extend is a placeholder.) The problem here comes from the comment above this code about the change in the catalog table's columns. I guess the only place to test which column names are valid is inside `_fields_from_ecsv()`.

Reply:
Turns out `"skycoord.ra"` and `"skycoord.dec"` are valid. I thought they were actual column names instead of variables set at the head of the .ecsv file. I'll change code related to this `cat_fields` back to its previous state.